addons/sale/models/account_invoice.py

```diff
@@ -38,8 +38,7 @@ def _onchange_partner_shipping_id(self):
     def unlink(self):
         downpayment_lines = self.mapped('invoice_line_ids.sale_line_ids').filtered(lambda line: line.is_downpayment)
         res = super(AccountInvoice, self).unlink()
-        if downpayment_lines:
-            downpayment_lines.unlink()
+        downpayment_lines.filtered(lambda line: not line.invoice_lines).unlink()
```

Comment:
You could filter directly in the first line of the method.

Reply:
yes, it should be better, I tried the other way around because it felt more intuitive to get what the second filter was looking for. But the double filter is just ugly and the unlink is actually called on the empty recordset, so I'd stick with the suggestion :)